core/services/offchainreporting/OCRContract.go

```diff
@@ -0,0 +1,232 @@
+	}
+	if len(logs) == 0 {
+		return c, errors.Errorf("ConfigFromLogs: OCRContract with address 0x%x has no logs", oc.contractAddress)
+	}
+
+	latest, err := oc.contractFilterer.ParseConfigSet(logs[len(logs)-1])
+	if err != nil {
+		return c, errors.Wrap(err, "ConfigFromLogs failed to ParseConfigSet")
+	}
+	latest.Raw = logs[len(logs)-1]
+	return confighelper.ContractConfigFromConfigSetEvent(*latest), err
+}
+
+func (oc *OCRContract) LatestBlockHeight(ctx context.Context) (blockheight uint64, err error) {
+	h, err := oc.ethClient.HeaderByNumber(ctx, nil)
+	if err != nil {
+		return 0, err
+	}
+	if h == nil {
+		return 0, errors.New("got nil head")
+	}
+
+	return uint64(h.Number), nil
+}
+
+func (oc *OCRContract) LatestTransmissionDetails(ctx context.Context) (configDigest ocrtypes.ConfigDigest, epoch uint32, round uint8, latestAnswer ocrtypes.Observation, latestTimestamp time.Time, err error) {
+	opts := bind.CallOpts{Context: ctx, Pending: false}
+	result, err := oc.contractCaller.LatestTransmissionDetails(&opts)
+	if err != nil {
+		return configDigest, 0, 0, ocrtypes.Observation(nil), time.Time{}, errors.Wrap(err, "error getting LatestTransmissionDetails")
+	}
+	return result.ConfigDigest, result.Epoch, result.Round, ocrtypes.Observation(result.LatestAnswer), time.Unix(int64(result.LatestTimestamp), 0), nil
+}
+
+func (oc *OCRContract) FromAddress() gethCommon.Address {
+	return oc.transmitter.FromAddress()
+}
+
+type OCRContractConfigSubscription struct {
+	logger          logger.Logger
+	contractAddress gethCommon.Address
+	ch              chan ocrtypes.ContractConfig
+	oc              *OCRContract
+	mutex           sync.Mutex
+	chClosed        bool
+}
+
+func (sub *OCRContractConfigSubscription) OnConnect() {}
+func (sub *OCRContractConfigSubscription) OnDisconnect() {
+	sub.mutex.Lock()
+	defer sub.mutex.Unlock()
+
+	if !sub.chClosed {
+		sub.chClosed = true
+		close(sub.ch)
+	}
+}
+func (sub *OCRContractConfigSubscription) HandleLog(lb eth.LogBroadcast, err error) {
+	topics := lb.Log().RawLog().Topics
+	if len(topics) == 0 {
+		return
+	}
+	switch topics[0] {
+	case OCRContractConfigSet:
+		raw := lb.Log().RawLog()
+		if raw.Address != sub.contractAddress {
+			sub.logger.Errorf("log address of 0x%x does not match configured contract address of 0x%x", raw.Address, sub.contractAddress)
+			return
+		}
+		configSet, err := sub.oc.contractFilterer.ParseConfigSet(raw)
+		if err != nil {
+			sub.logger.Errorw("could not parse config set", "err", err)
+			return
+		}
+		configSet.Raw = lb.Log().RawLog()
+		cc := confighelper.ContractConfigFromConfigSetEvent(*configSet)
+		sub.ch <- cc
+	default:
+	}
+}
+
+func (sub *OCRContractConfigSubscription) JobID() *models.ID {
+	jobID := sub.oc.jobID
+	return &jobID
+}
+
+func (sub *OCRContractConfigSubscription) Configs() <-chan ocrtypes.ContractConfig {
+	return sub.ch
+}
+
+func (sub *OCRContractConfigSubscription) Close() {
+	sub.oc.logBroadcaster.Unregister(sub.oc.contractAddress, sub)
```

Comment:
Re: my above comment about `OnDisconnect`, this looks like the place where you'd want to close the channel.